org.eclipse.jdt.core/codeassist/org/eclipse/jdt/internal/codeassist/CompletionEngine.java

```diff
@@ -12083,135 +12083,157 @@ private void findTypesFromExpectedTypes(char[] token, Scope scope, ObjectVector
 
 				checkCancel();
 
+				ReferenceBinding refBinding;
+				boolean receiverIsArrayType = false;
 				if(this.expectedTypes[i] instanceof ReferenceBinding) {
-					ReferenceBinding refBinding = (ReferenceBinding)this.expectedTypes[i];
+					refBinding = (ReferenceBinding) this.expectedTypes[i];
+				} else if (this.expectedTypes[i] instanceof ArrayBinding) {
+					receiverIsArrayType = true;
+					TypeBinding leaf = ((ArrayBinding) this.expectedTypes[i]).leafComponentType();
+					refBinding = this.lookupEnvironment
+							.getType(CharOperation.splitOn('.', CharOperation.concat(leaf.qualifiedPackageName(),
+									leaf.qualifiedSourceName(), '.')));
```

Comment:
I'm not totally sure what's going on here:
* why the detour from an existing binding via its name to a new binding?
* what if `refBinding` is not a valid type binding afterwards?
* I'm missing an explicit check for array-of-primitive, perhaps that's what the detour achieves in an obscure way? :)

Reply:
I will check this again @stephan-herrmann and update. I think I have to refresh my approach here :).